Some block processing unit tests force processing of inconsistent blocks

`test_on_block_outside_safe_slots_and_multiple_better_justified` and `test_on_block_outside_safe_slots_but_finality` in `test/phase0/unittests/fork_choice/test_on_block.py` force the processing of inconsistent blocks. 

## Problem:
Context:
```python
    # Mock justified block in store
    just_block = build_empty_block_for_next_slot(spec, state)
    # Slot is same as justified checkpoint so does not trigger an override in the store
    just_block.slot = spec.compute_start_slot_at_epoch(store.justified_checkpoint.epoch)
    store.blocks[just_block.hash_tree_root()] = just_block
```
The inconsistency is caused when `just_block.slot` is being set - the new slot is earlier than `just_block`'s parent block. 

Here[[1](https://github.com/ethereum/eth2.0-specs/blame/9eb662786adebce7c3e41852a525458895e8bedc/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py#L237-L238), [2](https://github.com/ethereum/eth2.0-specs/blame/9eb662786adebce7c3e41852a525458895e8bedc/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py#L290-L291)] is an illustration of the problem:
```python
    just_block_parent = store.blocks[just_block.parent_root]
    assert just_block_parent.slot < just_block.slot, f"just_block_parent.slot: {just_block_parent.slot}, just_block.slot: {just_block.slot}"
```

Both unit tests now fail with this output:
```
AssertionError: just_block_parent.slot: 16, just_block.slot: 0
```

This is fine in the current spec, because only the block's slot value matters for these unit tests. However, it causes a problem for a planned fork choice upgrade where constructing the block tree becomes relevant for this unit test.

## Suggested Fix:
Create consistent block hierarchy for `just_block`.

--- a/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
+++ b/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
@@ -273,7 +273,7 @@ def test_on_block_outside_safe_slots_and_multiple_better_justified(spec, state):
 def test_on_block_outside_safe_slots_but_finality(spec, state):
     # Initialization
     store = get_genesis_forkchoice_store(spec, state)
-    time = 0
+    time = 100
     spec.on_tick(store, time)
 
     next_epoch(spec, state)

--- a/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
+++ b/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
@@ -239,9 +239,6 @@ def test_on_block_outside_safe_slots_and_multiple_better_justified(spec, state):
     just_block = build_empty_block_for_next_slot(spec, state)
     store.blocks[just_block.hash_tree_root()] = just_block
 
-    just_block_parent = store.blocks[just_block.parent_root]
-    assert just_block_parent.slot < just_block.slot, f"just_block_parent.slot: {just_block_parent.slot}, just_block.slot: {just_block.slot}"
-
     # Step time past safe slots
     spec.on_tick(store, store.time + spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED * spec.SECONDS_PER_SLOT)
     assert spec.get_current_slot(store) % spec.SLOTS_PER_EPOCH >= spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED
@@ -296,9 +293,6 @@ def test_on_block_outside_safe_slots_but_finality(spec, state):
     # Create new higher justified checkpoint not in branch of store's justified checkpoint
     just_block = build_empty_block_for_next_slot(spec, state)
     store.blocks[just_block.hash_tree_root()] = just_block
-
-    just_block_parent = store.blocks[just_block.parent_root]
-    assert just_block_parent.slot < just_block.slot, f"just_block_parent.slot: {just_block_parent.slot}, just_block.slot: {just_block.slot}"
 
     # Step time past safe slots
     spec.on_tick(store, store.time + spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED * spec.SECONDS_PER_SLOT)

--- a/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
+++ b/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
@@ -225,13 +225,13 @@ def test_on_block_outside_safe_slots_and_multiple_better_justified(spec, state):
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
     state, store, last_signed_block = apply_next_epoch_with_attestations(spec, state, store)
     last_block_root = hash_tree_root(last_signed_block.message)
-    
+
     # Mock ficticious justified checkpoint in store
     store.justified_checkpoint = spec.Checkpoint(
         epoch=spec.compute_epoch_at_slot(last_signed_block.message.slot),
         root=spec.Root("0x4a55535449464945440000000000000000000000000000000000000000000000")
     )
-    
+
     next_epoch(spec, state)
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
 
@@ -280,13 +280,13 @@ def test_on_block_outside_safe_slots_but_finality(spec, state):
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
     state, store, last_signed_block = apply_next_epoch_with_attestations(spec, state, store)
     last_block_root = hash_tree_root(last_signed_block.message)
-    
+
     # Mock ficticious justified checkpoint in store
     store.justified_checkpoint = spec.Checkpoint(
         epoch=spec.compute_epoch_at_slot(last_signed_block.message.slot),
         root=spec.Root("0x4a55535449464945440000000000000000000000000000000000000000000000")
     )
-    
+
     next_epoch(spec, state)
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
 

--- a/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
+++ b/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
@@ -234,6 +234,9 @@ def test_on_block_outside_safe_slots_and_multiple_better_justified(spec, state):
     just_block.slot = spec.compute_start_slot_at_epoch(store.justified_checkpoint.epoch)
     store.blocks[just_block.hash_tree_root()] = just_block
 
+    just_block_parent = store.blocks[just_block.parent_root]
+    assert just_block_parent.slot < just_block.slot, f"just_block_parent.slot: {just_block_parent.slot}, just_block.slot: {just_block.slot}"
+
     # Step time past safe slots
     spec.on_tick(store, store.time + spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED * spec.SECONDS_PER_SLOT)
     assert spec.get_current_slot(store) % spec.SLOTS_PER_EPOCH >= spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED
@@ -283,6 +286,9 @@ def test_on_block_outside_safe_slots_but_finality(spec, state):
     # Slot is same as justified checkpoint so does not trigger an override in the store
     just_block.slot = spec.compute_start_slot_at_epoch(store.justified_checkpoint.epoch)
     store.blocks[just_block.hash_tree_root()] = just_block
+
+    just_block_parent = store.blocks[just_block.parent_root]
+    assert just_block_parent.slot < just_block.slot, f"just_block_parent.slot: {just_block_parent.slot}, just_block.slot: {just_block.slot}"
 
     # Step time past safe slots
     spec.on_tick(store, store.time + spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED * spec.SECONDS_PER_SLOT)

--- a/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
+++ b/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
@@ -307,13 +307,15 @@ def test_on_block_outside_safe_slots_but_finality(spec, state):
     # Mock justified and finalized update in state
     just_fin_state = store.block_states[last_block_root]
     new_justified = spec.Checkpoint(
-        epoch=store.justified_checkpoint.epoch + 1,
+        epoch=spec.compute_epoch_at_slot(just_block.slot) + 1,
         root=just_block.hash_tree_root(),
     )
+    assert new_justified.epoch > store.justified_checkpoint.epoch
     new_finalized = spec.Checkpoint(
-        epoch=store.finalized_checkpoint.epoch + 1,
+        epoch=spec.compute_epoch_at_slot(just_block.slot),
         root=just_block.parent_root,
     )
+    assert new_finalized.epoch > store.finalized_checkpoint.epoch
     just_fin_state.current_justified_checkpoint = new_justified
     just_fin_state.finalized_checkpoint = new_finalized
 

--- a/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
+++ b/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
@@ -226,7 +226,7 @@ def test_on_block_outside_safe_slots_and_multiple_better_justified(spec, state):
     state, store, last_signed_block = apply_next_epoch_with_attestations(spec, state, store)
     last_block_root = hash_tree_root(last_signed_block.message)
 
-    # Mock ficticious justified checkpoint in store
+    # Mock fictitious justified checkpoint in store
     store.justified_checkpoint = spec.Checkpoint(
         epoch=spec.compute_epoch_at_slot(last_signed_block.message.slot),
         root=spec.Root("0x4a55535449464945440000000000000000000000000000000000000000000000")
@@ -281,7 +281,7 @@ def test_on_block_outside_safe_slots_but_finality(spec, state):
     state, store, last_signed_block = apply_next_epoch_with_attestations(spec, state, store)
     last_block_root = hash_tree_root(last_signed_block.message)
 
-    # Mock ficticious justified checkpoint in store
+    # Mock fictitious justified checkpoint in store
     store.justified_checkpoint = spec.Checkpoint(
         epoch=spec.compute_epoch_at_slot(last_signed_block.message.slot),
         root=spec.Root("0x4a55535449464945440000000000000000000000000000000000000000000000")

--- a/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
+++ b/tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_block.py
@@ -224,14 +224,19 @@ def test_on_block_outside_safe_slots_and_multiple_better_justified(spec, state):
     next_epoch(spec, state)
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
     state, store, last_signed_block = apply_next_epoch_with_attestations(spec, state, store)
+    last_block_root = hash_tree_root(last_signed_block.message)
+    
+    # Mock ficticious justified checkpoint in store
+    store.justified_checkpoint = spec.Checkpoint(
+        epoch=spec.compute_epoch_at_slot(last_signed_block.message.slot),
+        root=spec.Root("0x4a55535449464945440000000000000000000000000000000000000000000000")
+    )
+    
     next_epoch(spec, state)
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
-    last_block_root = hash_tree_root(last_signed_block.message)
 
-    # Mock justified block in store
+    # Create new higher justified checkpoint not in branch of store's justified checkpoint
     just_block = build_empty_block_for_next_slot(spec, state)
-    # Slot is same as justified checkpoint so does not trigger an override in the store
-    just_block.slot = spec.compute_start_slot_at_epoch(store.justified_checkpoint.epoch)
     store.blocks[just_block.hash_tree_root()] = just_block
 
     just_block_parent = store.blocks[just_block.parent_root]
@@ -271,20 +276,25 @@ def test_on_block_outside_safe_slots_and_multiple_better_justified(spec, state):
 def test_on_block_outside_safe_slots_but_finality(spec, state):
     # Initialization
     store = get_genesis_forkchoice_store(spec, state)
-    time = 100
+    time = 0
     spec.on_tick(store, time)
 
     next_epoch(spec, state)
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
     state, store, last_signed_block = apply_next_epoch_with_attestations(spec, state, store)
+    last_block_root = hash_tree_root(last_signed_block.message)
+    
+    # Mock ficticious justified checkpoint in store
+    store.justified_checkpoint = spec.Checkpoint(
+        epoch=spec.compute_epoch_at_slot(last_signed_block.message.slot),
+        root=spec.Root("0x4a55535449464945440000000000000000000000000000000000000000000000")
+    )
+    
     next_epoch(spec, state)
     spec.on_tick(store, store.time + state.slot * spec.SECONDS_PER_SLOT)
-    last_block_root = hash_tree_root(last_signed_block.message)
 
-    # Mock justified block in store
+    # Create new higher justified checkpoint not in branch of store's justified checkpoint
     just_block = build_empty_block_for_next_slot(spec, state)
-    # Slot is same as justified checkpoint so does not trigger an override in the store
-    just_block.slot = spec.compute_start_slot_at_epoch(store.justified_checkpoint.epoch)
     store.blocks[just_block.hash_tree_root()] = just_block
 
     just_block_parent = store.blocks[just_block.parent_root]

